libcloudforensics/providers/aws/internal/ec2.py

```diff
@@ -137,3 +142,368 @@ def AttachVolume(self,
+    """List AMI images.
+
+    Args:
+      qfilter (List[Dict]): The filter expression.
+      See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_images  # pylint: disable=line-too-long
+
+    Returns:
+      List[Dict[str, Any]]: The list of images with their properties.
+
+    Raises:
+      RuntimeError: If the images could not be listed.
+    """
+    if not qfilter:
+      qfilter = []
+
+    client = self.aws_account.ClientApi(common.EC2_SERVICE)
+    try:
+      images = client.describe_images(
+          Filters=qfilter)  # type: Dict[str, List[Dict[str, Any]]]
+    except client.exceptions.ClientError as exception:
+      raise RuntimeError(str(exception))
+
+    return images['Images']
+
+  def GetOrCreateAnalysisVm(
+      self,
+      vm_name: str,
+      boot_volume_size: int,
+      ami: str,
+      cpu_cores: int,
+      boot_volume_type: str = 'gp2',
+      packages: Optional[List[str]] = None,
+      ssh_key_name: Optional[str] = None,
+      tags: Optional[Dict[str, str]] = None) -> Tuple[AWSInstance, bool]:
+    """Get or create a new virtual machine for analysis purposes.
+
+    Args:
+      vm_name (str): The instance name tag of the virtual machine.
+      boot_volume_size (int): The size of the analysis VM boot volume (in GB).
+      ami (str): The Amazon Machine Image ID to use to create the VM.
+      cpu_cores (int): Number of CPU cores for the analysis VM.
+      boot_volume_type (str): Optional. The volume type for the boot volume
+          of the VM. Can be one of 'standard'|'io1'|'gp2'|'sc1'|'st1'. The
+          default is 'gp2'.
+      packages (List[str]): Optional. List of packages to install in the VM.
+      ssh_key_name (str): Optional. A SSH key pair name linked to the AWS
+          account to associate with the VM. If none provided, the VM can only
+          be accessed through in-browser SSH from the AWS management console
+          with the EC2 client connection package (ec2-instance-connect). Note
+          that if this package fails to install on the target VM, then the VM
+          will not be accessible. It is therefore recommended to fill in this
+          parameter.
+      tags (Dict[str, str]): Optional. A dictionary of tags to add to the
+          instance, for example {'TicketID': 'xxx'}. An entry for the instance
+          name is added by default.
+
+    Returns:
+      Tuple[AWSInstance, bool]: A tuple with an AWSInstance object and a
+          boolean indicating if the virtual machine was created (True) or
+          reused (False).
+
+    Raises:
+      RuntimeError: If the virtual machine cannot be found or created.
+    """
+
+    # Re-use instance if it already exists, or create a new one.
+    try:
+      instances = self.GetInstancesByName(vm_name)
+      if instances:
+        created = False
+        return instances[0], created
+    except RuntimeError:
```

Comment:
We desperately need more specific error types here

Reply:
This will be taken care of in #223